primitives/src/json.rs

```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+impl serde::ser::Serialize for JsonString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let json: Value = serde_json::from_str(&self.0).map_err(serde::ser::Error::custom)?;
+        json.serialize(serializer)
+    }
+}
+
+impl iroha_schema::TypeId for JsonString {
+    fn id() -> iroha_schema::Ident {
+        "JsonString".to_owned()
+    }
+}
+
+impl IntoSchema for JsonString {
+    fn type_name() -> iroha_schema::Ident {
+        <Self as iroha_schema::TypeId>::id()
+    }
+
+    fn update_schema_map(map: &mut iroha_schema::MetaMap) {
+        if !map.contains_key::<Self>() {
+            map.insert::<Self>(iroha_schema::Metadata::String);
+        }
+    }
+}
+
+impl<T: Serialize> TryFrom<JsonCandidate<T>> for JsonString {
+    type Error = serde_json::Error;
+    fn try_from(value: JsonCandidate<T>) -> Result<Self, Self::Error> {
+        Ok(JsonString(serde_json::to_string(&value.0)?))
+    }
+}
+
+impl PartialOrd for JsonString {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for JsonString {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        let l: Value = serde_json::from_str(&self.0).unwrap();
+        let r: Value = serde_json::from_str(&other.0).unwrap();
+
+        match (l, r) {
+            (Value::Number(left), Value::Number(right)) => {
+                if left.is_u64() && right.is_u64() {
+                    left.as_u64().cmp(&right.as_u64())
+                } else {
+                    left.as_f64().unwrap().total_cmp(&right.as_f64().unwrap())
+                }
+            }
+            (Value::Array(left), Value::Array(right)) => {
+                let left = left.iter().map(JsonString::from);
+                let right = right.iter().map(JsonString::from);
+                left.cmp(right)
+            }
+            (Value::Object(left), Value::Object(right)) => {
+                let left = left.iter().map(|(_, v)| JsonString::from(v));
+                let right = right.iter().map(|(_, v)| JsonString::from(v));
+                left.cmp(right)
+            }
+            _ => self.0.cmp(&other.0),
+        }
+    }
+}
+
+impl From<&Value> for JsonString {
+    fn from(value: &Value) -> Self {
+        JsonString(value.to_string())
+    }
+}
+
+impl From<Value> for JsonString {
+    fn from(value: Value) -> Self {
+        JsonString(value.to_string())
+    }
+}
+
+impl From<u32> for JsonString {
+    fn from(value: u32) -> Self {
+        JsonString::new(value)
+    }
+}
+
+impl From<u64> for JsonString {
+    fn from(value: u64) -> Self {
+        JsonString::new(value)
+    }
+}
+
+impl From<f64> for JsonString {
+    fn from(value: f64) -> Self {
+        JsonString::new(value)
+    }
+}
+
+impl From<bool> for JsonString {
+    fn from(value: bool) -> Self {
+        JsonString::new(value)
+    }
+}
```

Comment:
Can't we use single implementation instead?

This way we can avoid panicking constructor.

```
impl<T: Serialize> From<T> for JsonString {
    // ...
}
```

Reply:
> Can't we use single implementation instead?
> 
> This way we can avoid panicking constructor.
> 
> ```
> impl<T: Serialize> From<T> for JsonString {
>     // ...
> }
> ```

No, we can not. It's prohibited to implement generic because of the trait implementation conflict

Reply:
Which traits are in conflict?

Reply:
Conflicting with `impl From<JsonString> for JsonString`... because `JsonString` itself implement `Serialize`